perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_groupNames.java

```diff
@@ -60,36 +58,35 @@ public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDe
 	}
 
 	@Override
-	public List<String> resolveVirtualAttributeValueChange(PerunSessionImpl sess, String message) throws InternalErrorException, AttributeNotExistsException, WrongAttributeAssignmentException {
-		List<String> resolvingMessages = new ArrayList<>();
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl sess, AuditEvent message) throws InternalErrorException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
 		if (message == null) return resolvingMessages;
 
-		Matcher memberAddedToMatcher = memberAddedToPattern.matcher(message);
-		Matcher memberTotallyRemovedFromMatcher = memberTotallyRemovedFromPattern.matcher(message);
+		if (message instanceof DirectMemberAddedToGroup) {
+			resolvingMessages.addAll(resolveDirectMemberAddedToGroup(sess, (DirectMemberAddedToGroup) message));
+		} else if (message instanceof MemberRemovedFromGroupTotally) {
+			resolvingMessages.addAll(resolveMemberRemovedFromGroupTotally(sess, (MemberRemovedFromGroupTotally) message));
+		}
+		return resolvingMessages;
+	}
 
-		User user;
-		Attribute attribute;
+	private List<AuditEvent> resolveDirectMemberAddedToGroup (PerunSessionImpl sess, DirectMemberAddedToGroup directMemberAddedToGroup) throws InternalErrorException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
 
-		if (memberAddedToMatcher.find() || memberTotallyRemovedFromMatcher.find()) {
+		User user = sess.getPerunBl().getUsersManagerBl().getUserByMember(sess, directMemberAddedToGroup.getMember());
+		Attribute attribute = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, user, WHOLE_NAME);
+		resolvingMessages.add(new AttributeSetForUser(attribute, user));
```

Comment:
Again the same problem. When you receive the attribute, you should check its value. If the value is null or empty, you should return event AttributeRemovedForMember. If the value is non-null and not empty, you should return AttributeSetForUser.